File: MODBUS_HID/Modbus.py
import hid
import struct
import time
from Modbus_def import *

def print_bytearray(bytearray):
    hex_str = ' '.join(f'{byte:02x}' for byte in bytearray)
    print(hex_str)

def format_command(command):
    bytearray_command = bytearray()
    for U16 in command:
        bytearray_command.extend([U16 >> 8, U16 & 0xFF])
        # print_bytearray(bytearray_command)
    return bytearray_command

def find_device(vendor_id, product_id):
    for device_dict in hid.enumerate():
        if device_dict['vendor_id'] == vendor_id and device_dict['product_id'] == product_id:
            return device_dict
    return None

def MasterModBus(device, FunCode, pIn, pOut):
    
    print(type(pIn))

    HidBuf = bytearray(HID_PACK_MAX + 1)
    HidBuf[HID_PACK_CH] = REPORT_ID
    HidBuf[HID_PACK_TYPE] = MB_HID_PROTOCOL_ID
    HidBuf[HID_PACK_MODBUS] = FunCode
    
    # Set packet length based on function code
    if FunCode == MBFunctions.MB_FUNC_READ_INPUT_REGISTER or FunCode == MBFunctions.MB_FUNC_READ_HOLDING_REGISTER:
        HidBuf[HID_PACK_LEN] = 7
    elif FunCode == MBFunctions.MB_FUNC_WRITE_MULTIPLE_REGISTERS:
        HidBuf[HID_PACK_LEN] = 7 + (pIn[4] + 1)
        if HidBuf[3] > HID_PACK_MAX:
            return MBErrorCode.MB_ELEN
    else:
        return MBErrorCode.MB_EILLFUNCTION
    
    # print("Write: ", HidBuf)

    # Copy input data to HidBuf
    HidBuf[HID_PACK_MODBUS+1:HID_PACK_MODBUS+1+len(pIn)] = pIn[:]
    HidBuf = HidBuf[:HID_PACK_MAX + 1]

    # Write HID packet
    print("Write: ")
    print_bytearray(HidBuf)
    if not device.write(HidBuf):
        return MBErrorCode.MB_EIO

    # Read response
    time.sleep(TIME_OUT / 1000.0)  # Simple delay for demonstration, consider adjusting for your application
    HidBuf = device.read(HID_PACK_MAX + 1)
    #add ID 
    HidBuf.insert(0, 0x0)
    if not HidBuf:
        return MBErrorCode.MB_ETIMEDOUT
    
    print("Read: ")
    print_bytearray(HidBuf)

    # Validate response length
    if HidBuf[HID_PACK_LEN] > HID_PACK_MAX:
        return MBErrorCode.MB_ELEN
    
    if HidBuf[HID_PACK_MODBUS] == FunCode:
        if HidBuf[HID_PACK_MODBUS] == MBFunctions.MB_FUNC_READ_INPUT_REGISTER or MBFunctions.MB_FUNC_READ_HOLDING_REGISTER: # Modbus function 0x04 Read Input Registers
            if((HidBuf[HID_PACK_LEN] != HidBuf[HID_PACK_MODBUS+1]+4) or (HidBuf[HID_PACK_LEN] & 0x01)):                     # Modbus function 0x03 Read Holding Registers
                return MBErrorCode.MB_ELEN
            else:
                for i in range(0, HidBuf[HID_PACK_MODBUS+1], 2):
                    pOut[i] = HidBuf[HID_PACK_MODBUS + 2 + i + 1]
                    pOut[i + 1] = HidBuf[HID_PACK_MODBUS + 2 + i]
            
        elif HidBuf[HID_PACK_MODBUS] == MBFunctions.MB_FUNC_WRITE_MULTIPLE_REGISTERS:                                       # Modbus function 0x10 Write Multiple Registers
            HidBuf[HID_PACK_LEN] = 0+5+(HidBuf[HID_PACK_MODBUS+5]*2+1)	

    elif HidBuf[HID_PACK_MODBUS] == (FunCode | 0x80):
        return HidBuf[HID_PACK_MODBUS+1] # Return exception code
    else:
        return MBErrorCode.MB_ERETURN
    
    return MBErrorCode.MB_EOK

def MasterRead(ReadType, RegStart, RegCount):
   
    pOut = bytearray()
    FunCode = MBFunctions.MB_FUNC_READ_HOLDING_REGISTER if ReadType == 0 else MBFunctions.MB_FUNC_READ_INPUT_REGISTER
    
    total_bytes = RegCount * 2  # Assuming each register is 2 bytes
    read_bytes = 0

    while RegCount > 0:
        current_read_count = min(RegCount, READ_REG_COUNT_MAX)
        RegCount -= current_read_count
        
        InBuf = bytearray([
            (RegStart >> 8) & 0xFF,  # High byte of RegStart
            RegStart & 0xFF,         # Low byte of RegStart
            0x00,                     # High byte of the register count (always 0 since count is <= READ_REG_COUNT_MAX)
            current_read_count       # Low byte of the register count
        ])
        
        ret = MasterModBus(device, FunCode, InBuf, pOut)
        if ret != MBErrorCode.MB_EOK:
            return ret, pOut[:total_bytes]
        
        # Assuming MasterModBus appends the read data to a global `pOut`
        # You might need to adjust this part depending on how MasterModBus is implemented
        read_bytes += current_read_count * 2
        RegStart += current_read_count

    # At this point, `pOut` should contain the data read
    # Depending on your implementation, you might need to return or process `pOut` further
    return MBErrorCode.MB_EOK, pOut[:total_bytes]

def MasterWrite(RegStart, RegCount, pIn):

    pOut = bytearray()
    # total_bytes_to_write = RegCount * 2  # Assuming 2 bytes per register
    written_bytes = 0  # Counter for the bytes written

    while RegCount > 0:
        current_write_count = min(RegCount, WRITE_REG_COUNT_MAX)
        RegCount -= current_write_count

        # Preparing InBuf with the register start address and the count of registers to write
        InBuf = bytearray([
            (RegStart >> 8) & 0xFF,  # High byte of RegStart
            RegStart & 0xFF,         # Low byte of RegStart
            0x00,                     # High byte of the count (always 0 since count is <= WRITE_REG_COUNT_MAX)
            current_write_count,      # Low byte of the count
            2 * current_write_count   # Number of bytes to write
        ])

        # Add the data to be written, swapping bytes
        for i in range(0, 2 * (current_write_count), 2):
            # print("index", written_bytes + i)
            InBuf.extend([pIn[written_bytes + i], pIn[written_bytes + i + 1]])
            
        # Send the write command
        ret = MasterModBus(device, MBFunctions.MB_FUNC_WRITE_MULTIPLE_REGISTERS, InBuf, pOut)
        if ret != MBErrorCode.MB_EOK:
            return ret

        written_bytes += 2 * current_write_count
        RegStart += current_write_count

    return MBErrorCode.MB_EOK

if __name__ == "__main__":
    device_info = find_device(VENDOR_ID, PRODUCT_ID)
    
    if device_info:
        print("Device found. Opening...")
        device = hid.device()
        device.open(VENDOR_ID, PRODUCT_ID)
        print("Device opened.")

        print("Writing registers...")
        command = [Operation.ONLY_BALANCE, Select_memory.PROGRAM_0, Select_channel.CHANNEL_1, Order_lock.UNLOCK, Order.STOP]
        ret = MasterWrite(CONTROL_REGISTERS, len(command)//2, command)
        print("Writing return code: ", ret)
    
        device.close()
    else:
        print("Device not found.")
        
    print("Goodbye!")

File: modbus/src/modbus.cpp
#include "modbus.hpp"
#include <iostream>
#include <stdio.h>
#include <cstdint>

#include <wchar.h>
#include "../../hidapi/hidapi/hidapi.h"
// #include "hidapi.h"

using namespace std;

#define READ_REG_COUNT_MAX		((HID_PACK_MAX-4)/2) //30
#define WRITE_REG_COUNT_MAX		((HID_PACK_MAX-8)/2) //28
#define TIME_OUT	500

//  gcc -fPIC -shared -o hidapi.so windows/hid.c -Ihidapi
//  g++ -fPIC -shared -o CModbus.so modbus.cpp -Ihidapi/hidapi -Lhidapi\build\src\windows -Ihidapi


// Need to implement the same thing with the hid api lib
#include <iomanip>
void DisplayArrayInHex(const uint8_t* array, size_t length) {
    printf("Array in hex: [");
    for (size_t i = 0; i < length; ++i) {
        // Print each byte in hex format
        printf("%02x", static_cast<int>(array[i]));
        if (i < length - 1) {
            printf(", ");
        }
    }
    printf("]\n");
}


class hid_dev {
private:
    hid_device *handle;

public:
    hid_dev() {
        // Initialize the hidapi library
		printf("Hid init\n");
        hid_init();

        // Open the device using the VID, PID, and optionally the Serial number.
        handle = hid_open(0x0483, 0x5751, NULL);
        if (!handle) {
            printf("Unable to open device\n");
            hid_exit();
        }
    }

    ~hid_dev() {
        // Close the HID device and cleanup
        hid_close(handle);
        hid_exit();
    }

    bool read(uint8_t* buffer, size_t length) {
        int res = hid_read(handle, buffer, length);
        return (res >= 0);
    }

    bool write(const uint8_t* buffer, size_t length) {
        int res = hid_write(handle, buffer, length);
        return (res >= 0);
    }
};

hid_dev device;

eMBErrorCode MasterRead(uint8_t ReadType,uint32_t RegStart,uint32_t RegCount,uint8_t *pOut)
{
	eMBErrorCode ret;
	uint32_t i;
	uint8_t InBuf[16];
	uint8_t FunCode;

	if(ReadType==0)
		FunCode = MB_FUNC_READ_HOLDING_REGISTER;
	else
		FunCode = MB_FUNC_READ_INPUT_REGISTER;

	for(i=0;i<RegCount/READ_REG_COUNT_MAX;i++)
	{
		InBuf[0] = (uint8_t)(RegStart >> 8);
		InBuf[1] = (uint8_t)(RegStart & 0xff);
		InBuf[2] = 0;
		InBuf[3] = READ_REG_COUNT_MAX;
		ret = MasterModBus(FunCode,InBuf,pOut,TIME_OUT);
		if(ret != MB_EOK)return ret;
		RegStart += READ_REG_COUNT_MAX;
		pOut += (2*READ_REG_COUNT_MAX);
	}

	if(RegCount%READ_REG_COUNT_MAX)
	{
		InBuf[0] = (uint8_t)(RegStart >> 8);
		InBuf[1] = (uint8_t)(RegStart & 0xff);
		InBuf[2] = 0;
		InBuf[3] = (uint8_t) RegCount%READ_REG_COUNT_MAX;	
		ret = MasterModBus(FunCode,InBuf,pOut,TIME_OUT);
		if(ret != MB_EOK)return ret;
	}
	return ret;
}

eMBErrorCode MasterWrite(uint32_t RegStart,uint32_t RegCount,uint8_t *pIn)
{
	eMBErrorCode ret;
	uint32_t i,j;
	uint8_t InBuf[80];

	for(i=0;i<RegCount/WRITE_REG_COUNT_MAX;i++)
	{
		InBuf[0] = (uint8_t)(RegStart >> 8);
		InBuf[1] = (uint8_t)(RegStart & 0xff);
		InBuf[2] = 0;
		InBuf[3] = WRITE_REG_COUNT_MAX;
		InBuf[4] = 2*WRITE_REG_COUNT_MAX;

		for(j=0;j<InBuf[4];j=j+2)
		{
			InBuf[5+j] = pIn[j+1];
			InBuf[5+j+1] = pIn[j];
		}
		ret = MasterModBus(MB_FUNC_WRITE_MULTIPLE_REGISTERS,InBuf,NULL,TIME_OUT);
		if(ret != MB_EOK)return ret;
		RegStart += WRITE_REG_COUNT_MAX;
		pIn += (2*WRITE_REG_COUNT_MAX);
	}

	if(RegCount%WRITE_REG_COUNT_MAX)
	{
		InBuf[0] = (uint8_t)(RegStart >> 8);
		InBuf[1] = (uint8_t)(RegStart & 0xff);
		InBuf[2] = 0;
		InBuf[3] = (uint8_t) RegCount%WRITE_REG_COUNT_MAX;
		InBuf[4] = 2*InBuf[3];

		// printf("MasterWrite\n");
		// DisplayArrayInHex(InBuf, HID_PACK_MAX+1);

		for(j=0;j<InBuf[4];j=j+2)
		{
			InBuf[5+j] = pIn[j];
			InBuf[5+j+1] = pIn[j+1];
		}
		ret = MasterModBus(MB_FUNC_WRITE_MULTIPLE_REGISTERS,InBuf,NULL,TIME_OUT);
		if(ret != MB_EOK)return ret;
	}
	return ret;
}

eMBErrorCode MasterModBus(uint8_t FunCode,uint8_t *pIn,uint8_t *pOut,uint32_t ms)
{
	int i;
	uint8_t HidBuf[HID_PACK_MAX+1];

	HidBuf[HID_PACK_CH] = REPORT_ID;
	HidBuf[HID_PACK_TYPE] = MB_HID_PROTOCOL_ID;
	HidBuf[HID_PACK_MODBUS] = FunCode;

	// printf("MasterModBus\n");
	// DisplayArrayInHex(HidBuf, HID_PACK_MAX+1);

	switch(FunCode)
	{
	case MB_FUNC_READ_INPUT_REGISTER:	//Modbus function 0x04 Read Input Registers
		if(0)
			return MB_ENOREG;
		HidBuf[HID_PACK_LEN] = 7;
		break;
	case MB_FUNC_READ_HOLDING_REGISTER:  //Modbus function 0x03 Read Holding Registers
		if(0)
			return MB_ENOREG;
		HidBuf[HID_PACK_LEN] = 7;
		break;
	case MB_FUNC_WRITE_MULTIPLE_REGISTERS:	//Modbus function 0x10 Write Multiple Registers
		if(0)
			return MB_ENOREG;
		HidBuf[HID_PACK_LEN] = 7+(pIn[4]+1);
		if(HidBuf[HID_PACK_LEN] > HID_PACK_MAX)return MB_ELEN;
		break;
	default:
		return MB_EILLFUNCTION;
	}
	//copy from pD
	for(i=0;i<HidBuf[HID_PACK_LEN]-3;i++)
		HidBuf[HID_PACK_MODBUS+1+i]=pIn[i];

	//trans
	// if(JsHID.Write(HidBuf,HID_PACK_MAX+1)==FALSE)return MB_EIO;
	printf("MasterModBus write\n");
    DisplayArrayInHex(HidBuf, HID_PACK_MAX+1);
    if (!device.write(HidBuf, HID_PACK_MAX+1)) return MB_EIO;
	HidBuf[HID_PACK_CH]=REPORT_ID;
    if (!device.read(HidBuf, HID_PACK_MAX+1)) return MB_ETIMEDOUT;
    DisplayArrayInHex(HidBuf, HID_PACK_MAX+1);
	// if(JsHID.Read(HidBuf,HID_PACK_MAX+1,ms)==FALSE)return MB_ETIMEDOUT;
	if(HidBuf[HID_PACK_LEN] > HID_PACK_MAX)return MB_ELEN;
	

	if(HidBuf[HID_PACK_MODBUS] == FunCode)
	{
		switch(FunCode)
		{
		case MB_FUNC_READ_INPUT_REGISTER:	//Modbus function 0x04 Read Input Registers
		case MB_FUNC_READ_HOLDING_REGISTER:  //Modbus function 0x03 Read Holding Registers
			if((HidBuf[HID_PACK_LEN] != HidBuf[HID_PACK_MODBUS+1]+4) || (HidBuf[HID_PACK_LEN] & 0x01))return MB_ELEN;
			//copy to pOut
			for(i=0;i<HidBuf[HID_PACK_MODBUS+1];i=i+2)
			{
				pOut[i] = HidBuf[HID_PACK_MODBUS+2+i+1]; 
				pOut[i+1] = HidBuf[HID_PACK_MODBUS+2+i]; 
			}
			break;
		case MB_FUNC_WRITE_MULTIPLE_REGISTERS:	//Modbus function 0x10 Write Multiple Registers
			HidBuf[HID_PACK_LEN] = 0+5+(HidBuf[HID_PACK_MODBUS+5]*2+1);	
			break;
		}
	}
	else if(HidBuf[HID_PACK_MODBUS] == (FunCode | 0x80))
		return (eMBErrorCode)HidBuf[HID_PACK_MODBUS+1];
	else
		return MB_ERETURN;

	return MB_EOK;
}